app/code/Magento/InventoryCatalogAdminUi/Controller/Adminhtml/Source/BulkAssignPost.php

```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalogAdminUi\Controller\Adminhtml\Source;
+
+use Magento\Backend\App\Action;
+use Magento\Framework\Controller\ResultFactory;
+use Magento\Framework\Validation\ValidationException;
+use Magento\InventoryCatalogAdminUi\Model\BulkSessionProductsStorage;
+use Magento\InventoryCatalogApi\Api\BulkSourceAssignInterface;
+
+class BulkAssignPost extends Action
+{
+    /**
+     * @see _isAllowed()
+     */
+    const ADMIN_RESOURCE = 'Magento_Catalog::products';
+
+    /**
+     * @var BulkSessionProductsStorage
+     */
+    private $bulkSessionProductsStorage;
+
+    /**
+     * @var BulkSourceAssignInterface
+     */
+    private $bulkSourceAssign;
+
+    /**
+     * @param Action\Context $context
+     * @param BulkSourceAssignInterface $bulkSourceAssign
+     * @param BulkSessionProductsStorage $bulkSessionProductsStorage
+     * @SuppressWarnings(PHPMD.LongVariable)
+     */
+    public function __construct(
+        Action\Context $context,
+        BulkSourceAssignInterface $bulkSourceAssign,
+        BulkSessionProductsStorage $bulkSessionProductsStorage
+    ) {
+        parent::__construct($context);
+
+        $this->bulkSessionProductsStorage = $bulkSessionProductsStorage;
+        $this->bulkSourceAssign = $bulkSourceAssign;
+    }
+
+    /**
+     * @return \Magento\Framework\App\ResponseInterface|\Magento\Framework\Controller\ResultInterface
+     */
+    public function execute()
+    {
+        $sourceCodes = $this->getRequest()->getParam('sources', []);
+        $skus = $this->bulkSessionProductsStorage->getProductsSkus();
+
+        try {
+            $count = $this->bulkSourceAssign->execute($skus, $sourceCodes);
```

Comment:
Asynchronous Operation (processing) should be used here

Reply:
Al the module should be reworked for async ops, it will be my very next step.